src/backend/optimizer/path/allpaths.c

```diff
@@ -2125,15 +2125,25 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 			subplan = subquery_planner(cteroot->glob, subquery, cteroot, cte->cterecursive,
 									   tuple_fraction, &subroot, config);
 
-			cteplaninfo->shared_plan = prepare_plan_for_sharing(cteroot, subplan);
+			cteplaninfo->subplan = (subplan->flow->locustype != CdbLocusType_SegmentGeneral) ?
+			                       prepare_plan_for_sharing(cteroot, subplan) :
+			                       subplan;
```

Comment:
If we look at some code above for the case when `gp_cte_sharing` is off and true inlinining takes place, you will notice that
some restriction quals may be pushed down to cte in order to optimize the access to relations (like filter at the SeqScan stage, etc).
`subquery = push_down_restrict(root, rel, rte, rel->relid, subquery);`

Therefore, this patch causes suboptimal filtration. For example:
```
create table t1 (i int);
create table t_repl (i int, j int) distributed replicated;
```
Before patch `gp_cte_sharing = off`:
```
explain (costs off) with cte as(
select * from t_repl)
select * from cte a join (select * from cte b join t1 using (i)) x on a.i=x.j where x.j > 5 and a.j > random();
                                        QUERY PLAN                                        
------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)
   ->  Hash Join
         Hash Cond: (t_repl.j = t_repl_1.i)
         ->  Hash Join
               Hash Cond: (t1.i = t_repl.i)
               ->  Seq Scan on t1
               ->  Hash
                     ->  Seq Scan on t_repl
                           Filter: (j > 5)
         ->  Hash
               ->  Broadcast Motion 1:3  (slice1; segments: 1)
                     ->  Result
                           ->  Seq Scan on t_repl t_repl_1
                                 Filter: ((i > 5) AND ((j)::double precision > random()))
 Optimizer: Postgres query optimizer
(15 rows)
```
After patch `gp_cte_sharing=on`:
```
explain (costs off) with cte as(
select * from t_repl)
select * from cte a join (select * from cte b join t1 using (i)) x on a.i=x.j where x.j > 5 and a.j > random();
                                          QUERY PLAN                                          
----------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)
   ->  Hash Join
         Hash Cond: (b.j = a.i)
         ->  Hash Join
               Hash Cond: (b.i = t1.i)
               ->  Subquery Scan on b
                     Filter: (b.j > 5)
                     ->  Seq Scan on t_repl
               ->  Hash
                     ->  Seq Scan on t1
         ->  Hash
               ->  Broadcast Motion 1:3  (slice1; segments: 1)
                     ->  Result
                           ->  Subquery Scan on a
                                 Filter: ((a.i > 5) AND ((a.j)::double precision > random()))
                                 ->  Seq Scan on t_repl t_repl_1
 Optimizer: Postgres query optimizer
(17 rows)
```
As you can see, the Filter went above the SeqScan, which is suboptimal

Reply:
@KnightMurloc, what do you think? Probably this behaviour is ok in this case, because for gp_cte_sharing we don't know final locus in advance. Moreover gp_cte_sharing is disabled by default, and I guess suboptimalilty described above does not really matter.

Reply:
at the time of planning the second plan, we already know the locus and can plan it again with optimizations it will turn out a little better. it would be good if we could be replace the previously planned plan with a new one.

Reply:
if optimization is only a push down filter, can we save a pointer to this field of the plan and replace it with new one after re-planning second plan?

Reply:
After patch with gp_cte_sharing=off, everything is ok. And yep, if we make a decision to inline the plan, it becomes too late to optimize the filtration

Reply:
we can also do planning again with true inline. since we usually deal with long analytical queries, this may make sense.

Reply:
> if optimization is only a push down filter, can we save a pointer to this field of the plan and replace it with new one after re-planning second plan?

We should try to investigate that option. But, overall, I think doing something with that will be overhead.

Reply:
> How we can do that?

we planning a plan if it has a SegmentGeneral locus, we use the code of true inline (we can move it into a function), and then delete plan with SegmentGeneral locus. something like this

Reply:
I think that the main puprose of this patch is to make the perfomance robust and prevent potential deadlocks. Again, gp_cte_sharing is turned off by default and creating such overhead for this patch is not worth it. If user wants an optimal plan just turn it off. Replanning everything post factum is unnecessary, I guess.